### ISSUE TITLE
Weighted PageRank

### ISSUE THREAD

**Issue by decentralion** [sourcecred/sourcecred#476]
# Weighted PageRank Graphs

## Domain introduction

We have a heterogeneous graph `G` which represents contributions/contributors
in SourceCred, and connections between them.

It's heterogeneous in the sense that there are many different "types" of nodes
and edges. In the interest of concreteness, here is a sample of of node types
and edge types we have or expect to add.

### Node types:

  - *GitHub users* are identities we want to assign cred to
  - *GitHub issues* contain documentation, design, bug reports, etc
  - *GitHub pull requests* merge new code into the repository
  - *Git blobs* contain source code
  - *Spotlights* represent cred for contributions that don't otherwise fit
    into the data model, e.g. a spotlight for participtaing in an offline
    design discussion, or for giving a talk on SourceCred

### Edge types:

 - *Authorship edges* connect an issue or pull request with the users that
   created it
 - *MergedAs edges* connect a pull request to the git commit it created
 - *Import edges* connect a Git blob containing source code to the git blobs
   containing code that blob imports
 - *Spotlight edges* connect a Spotlight node to the entities it is flowing cred
   to
 - *Spotlight authorships* flow some cred to the entity that created the spotlight

### The need for weights

As currently implemented, SourceCred does not do a good job of distinguishing
valuable contributions. As an example, when run on the repository in its
current state, it believes that #292 is one of the most valuable pulls in the
project, outshining contributions like #252 and #468 which were actually much
more valuable. Even worse, it massively over-values Git objects; for example,
the blob containing our Prettier configuration is seen as >100x more valuable
than the most valuable pull request.

@wchargin and @decentralion believe that this problem results from the fact
that the graph is currently unweighted. As such, we need to implement a system
for applying weights to different nodes and to the src/dst of edges. There are
several different use cases for weights:

#### Prior over types

Weights can provide prior information on the relative value of different types.
For example, we should have a weight that suggests that GitHub pull requests
are, on average, much more important than Git blobs.

#### Domain-specific evaluation

Weights can use domain-specific signal to differentiate the value of different
nodes of the same type. For example, # of lines of code is a useful signal for
the value of a pull request; a PR that modifies 1000 lines of code is probably
more valuable than one that modifies 3 lines.

#### Distinguish parts of the project

Weights can be used to differentiate different parts of the project for clearer
analysis. For example, suppose a project wants to separately measure cred for
"backend" and "frontend" work, without artificially partioning the project into
separate graphs. It would be nice to do this via a "weighting" that assigns a
high prior weight to the backend directory and key backend APIs. (This would
incorporate the [project] or "domain" concept via weighting).

[project]:
https://github.com/sourcecred/sourcecred/issues/29#issuecomment-368359785

Thus, implementing a well-designed weighting abstraction would allow us to
implement many important features in a cohesive way. What follows is a "Q&A"
style discussion of major open questions in how the weight abstraction should
work.

### What is the type signature of a weight function?

In some cases, it's natural to think of a weight function as providing a
relative weight over every node in the project. For example, the project may
have maintainer-configurable weight functions that give a prior over every type
of node in te graph

In other cases, it's natural to think of a weight function as providing
relative weights for a subset of nodes in the graph. For example, the
lines-of-code weighting can differentiate between different pull request nodes,
but has no information to offer on the relative importance of different issues.

Some ideas on how we could express weight functions:

`node \in Graph -> number`, where for some subset of nodes in the graph, the
weight function assigns a weight. That weight might be seen as an update on the
a priori importance of the node. (What would the formal specification be?)

`Graph -> distribution over nodes`, where for every node in the graph, the
weight function assigns an update on the prior importance of that node. (Would
it assign 0 to every node that is out-of-scope? Would that compose nicely?)

### How do we compose weight functions?

It's important that we can combine and compose weight functions. For example,
suppose we have one weight function on commits that gives the logarithm of the
number of lines changed by the commit, and another weight function measures the
[cyclomatic complexity]. At the very least, we should be able to use both
weight functions in tandem.  It would be cool if we could express richer
relationships, such as that commits that add a lot of lines of code but do not
add any test code actually have negative signal.  (Unless they are refactors?)

cyclomatic complexity: [https://en.wikipedia.org/wiki/Cyclomatic_complexity]

We should also be able to compose weight functions "cross-product-style".
Suppose we have one pair of weight functions that distinguishes the frontend
and backend components, and we have another pair of weight functions that
distinguishes code implementation from documentation.

Then it should be possible to answer who has the most documentation cred in the
backend, who has the most implementation cred in the frontend, and so forth.

One insight from @krzhang was to think of the weight functions as priors in a
sequence of bayesian updates; he suggested that if we store the weights as
log-priors, then combining them is as simple as adding the log-priors.

### How do we apply the weight functions to the PageRank algorithm?

Classical PageRank doesn't provide prior weighting on nodes. One way to add
weighting, found in the weighting around [personalized PageRank], is to add
[personalization vectors].  This modifies the "random surfing" or
"teleportation" of the PageRank algorithm so that rather than uniformly
teleporting to any node in the graph, the surfer teleports based on the
personalization vector.

[personalized PageRank]: https://nlp.stanford.edu/projects/pagerank.shtml
[personalization vectors]: http://ilpubs.stanford.edu:8090/596/1/2003-35.pdf

In offline discussion, @krzhang suggested a different approach. We can apply a
node weight by applying a weighting to the in-and-out edge weights for that
node; i.e., by modifying that node's row and column in the adjacency matrix.

He suggested that given an adjacency matrix `M`, this could be applied by `W *
M * W'`, where `W` is a diagonal matrix with node weights on the diagonal.

This approach seems more powerful and principled to me - easier to reason about
how it affects the final distribution. We may want to empirically try both approaches.

---

Let's use this thread to further discuss this model; eventually we'll likely want to commit the final document as a markdown or latex document.

Thanks to @krzhang for helping to clarify the problem domain, and providing a number of suggestions that informed this post.

**Comment by decentralion** [comment]
Here's a proposal on weight function semantics.

Let a weight function map from a subset of nodes in the graph to real numbers,
where for each node `n`, `W(n)` represents the logarithm of the relative
significance of that node. For example, suppose we have a graph with the
following nodes:

- `r`: SourceCred repository
- `p1`, `p2`: pull requests
- `i1`: an issue
- `c1`, `c2`, `c3`, `c4`: commits

One example of a weight function is `w`, a prior over node types, defined as
follows. (I use base two logarithms due to personal preference.)

```
w(r) = 1
w(p1, p2) = 0
w(i1) = -1
w(c1, c2, c3, c4) = -2
```

This expreses that repositories are twice as important as pulls, which are
twice as important as issues, which are twice as important as commits.
Obviously, this is equivalent to another weight function, `w'`, defined as
follows:

```
w'(r) = 0
w'(p1, p2) = -1
w'(i1) = -2
w'(c1, c2, c3, c4) = -3
```

Within this equivalence class, let's consider "canonical" the weight function
which sums to 0, which in this case is 

```
w_c(r) = 2
w_c(p1, p2) = 1
w_c(i1) =  0
w_c(c1, c2, c3, c4) = -1
```

Now let's consider another weight function, this time defined only over the pull requests.
Perhaps it is counting the lines of code, so we can call it `loc`.

```
loc(p1) = 1
loc(p2) = 0
```

This expresses that p1 is twice as important as p2. However, it is not
meaningfully defined over all of the nodes in the graph. It might seem natural
to pad the remaining entries with zero, but this makes equivalent relative
weightings of pulls no longer equivalent in how they modify the whole graph.
Consider the functions `loc'` and `loc''`

```

loc'(p1) = 1
loc'(p2) = 0
loc'(r, i1, c1, c2, c3, c4) = 0


loc''(p1) = 0
loc''(p2) = -1
loc''(r1, i1, c1, c2, c3, c4) = 0
```

Even though `loc'` and `loc''` are expressing the same relative weightings
between pull requests, `loc'` also shifts overall weight in the graph towards
pull requests in general, while `loc''` shifts overall weight in the graph away
from pull requests. This is an unwanted side effect. Therefore, I think it
makes more sense for weighting functions only to be defined for the subset of
nodes that they explicitly compare. We can then define a function
`normalize` which turns the weighting into a canonical weight distribution via
the following procedure:

```javascript
type LogWeight = number;
// Not guaranteed to have every node in the graph. Not normalized
type Weighting = Map<NodeAddressT, LogWeight>;

// Guaranteed to have every node in the graph, and is normalized.
type NormalizedWeighting = Map<NodeAddressT, LogWeight>;

normalize(graph: Graph, weighting: Weighting): NormalizedWeighting
  let totalWeight = 0;
  let numWeights = 0;
  for (const weight of weighting.values()) {
    totalWeight += weight;
    numWeights++;
  }
  const normalizer = totalWeight / numWeights;
  const normWeighting: NormalizedWeighting = new Map();
  for (const node of graph.nodes()) {
    const w = weighting.get(node);
    normWeighting.set(node, w == null ? 0 : w - normalizer)
  }
  return normWeighting;
```

This has the advantage that it makes it easy to specify relative comparisons between
different nodestypes in easy-to-grok units (a is 2x as important as b), without accidentally
shifting the distribution over nodes that are out of scope for that weighting.

To round things out, here are psuedocode implementations of the three types of
weight functions described above.


```javascript

### Prior over types

const typeToWeight = {
  GitHub.Repo: 5,
  GitHub.User: 3,
  GitHub.Pull: 2,
  GitHub.Issue: 1,
  GitHub.Review: 0,
  GitHub.Comment: -1,
  Git.Commit: 0,
  Git.Tree: -4,
  Git.Blob: -5,
  Git.TreeEntry: -5,
}

const typeWeight = (n: NodeAddressT) => {
  return typeToWeight[getType(n)]
}

### Domain specific comparison

const locWeight = (n: NodeAddressT) => {
  if (getType(n) === GitHub.Pull) {
    const pull = githubRelationalView.get(n);
    const weight = Math.log(pull.additions() + pull.deletions);
    return weight;
  }
  return undefined;
}


### Subproject comparison

const canonicalBackendNodes = new Set([
  // list of nodes, as voted on by users or created via TCR or 
  // some other mechanism
])

const backendWeight = (n: NodeAddressT) => {
  if (canonicalBackendNodes.has(n)) {
  if (canonicalBackendNodes.has(n)) {
    return 4; // tunable parameter
  } else {
    return 0;
  }
}
```

cc @wchargin

**Comment by wchargin** [comment]
We like the idea of log-space priors, because it’s a natural space in
which to think. But the semantics for composing them as described above
don’t have the properties that we want.

First, let’s make the above semantics a bit more precise. Assume that we
only consider finite, inhabited types. Then:

  - A `LogDist α` is a total map from `α` to `Real` whose image is
    centered around `0`.

  - A `Dist α` is a valid probability distribution on `α`, represented
    as a function `f : α → NonnegativeReal` with the constraint that
    `sum (values f) == 1`.

  - We have `extend : ∀ α, β. (α <: β) ⇒ LogDist α → LogDist β`, defined by

    ```
    extend f (x : α) = f x
    extend f _ = 0
    ```

  - We have `normalize : ∀ α. (α → NonnegativeReal) → Dist α` given by

    ```
    normalize f x = f x / total
      where total = sum (values f)
    ```

  - We have `reify : ∀ α. LogDist α → Dist α` given by

    ```
    reify f = normalize (exp ∘ f)
    ```

  - We would like to define a notion of combination of log-distributions:

    ```
    (<>) : ∀ α. LogDist α → LogDist α → LogDist α
    ```

  - This operation should have at least the following property:

    **Irrelevance.** If `f, g : LogDist α` and `x : α`, and `f x == 0`,
    then `reify g x == reify (f <> g) x`.
    
From this formulation, there are a few things to note:

  - The implementation of `(<>)` given simply by pointwise addition does
    not satisfy the irrelevance property. It will be true that if
    `f x == 0` then `g x == (f <> g) x`, but it need not be true that
    `reify g x == reify (f <> g) x`, because the normalization factors
    in `g` and `f <> g` are, in general, different.

  - The `Dist` and `LogDist` functors are isomorphic, and in particular
    the operator `(<>)` induces an operator `λ f. λ g. reify (f <> g)`
    for combining non-log-space distributions. The problems are clearly
    equivalent, and while it may be more convenient to work in one space
    or the other, any fundamental problems must exist in both spaces.

  - Although `(<>)` is a simple operation in log-space, it’s not clear
    how to interpret this in distribution space. We can write down what
    it does, but how, intuitively, is it actually transforming the
    values? What is the interpretation?

  - The `Dist` functor happens to be a very useful monad (in my opinion,
    an underappreciated one), which already has a notion of combination
    of distributions: `join : ∀ α. Dist (Dist α) → Dist α`. This
    necessarily requires a prior over the distributions themselves. We
    should keep this in mind while designing our semantics: if we
    _don’t_ require a prior over these distributions (as in the API for
    our `(<>)` above), then what are we doing? Are we assuming a uniform
    prior? Is that a good thing?

**Comment by wchargin** [comment]
Suppose that we have a row-stochastic weighted adjacency matrix `M`, and
a weighting vector `w`, and let `W = diag(w)` as above. The above post
suggests setting `M′ = W M W`. The result of this is that an edge from
`u` to `v` with change weight from `t` to `t w(u) w(v)`. The resulting
`M′` is not row-stochastic, so we have to re-normalize it, and at that
point the factor `w(u)` in its definition (i.e., the left-multiplication
by `W`) cancels out.

So, instead, we may simply define `M′ = row-normalize(M · diag(w))`.

It is useful to reiterate that this is equivalent to multiplying each
edge’s weight by `w(v)`, where `v` is the destination of the edge.

**Comment by wchargin** [comment]
(Not sure exactly where to put this, but wanted to write it down.)

When thinking about ways to compose heuristics, we’ve often found that
it would be helpful to assume that the nodes can be partitioned by some
notion of “type”, and that every heuristic assigns a score to either
every node of a given type or none of them. For instance, if nodes are
divided into “issues” and “pull requests”, then a heuristic must either
evaluate every pull request relative to the other pull requests, or
express no judgment about any pull request.

I’m not sure that this assumption is realistic. It seems natural that
partial heuristics will want to evaluate nontrivially intersecting sets
of nodes. For instance, suppose the we have a repository where every
node has a “kind” either issue or pull request, and every node has an
“area” either documentation, implementation, or community support. Then
one heuristic might judge pull request nodes based on numbers of lines
of code, while another heuristic might judge community support nodes
based on how many first-time users reacted with “:+1:”. This breaks the
assumption above, because “kind” and “area” are orthogonal.

**Comment by wchargin** [comment]
So, today I learned about Snorkel:
<https://hazyresearch.github.io/snorkel/>

As described to me by @000drax: Snorkel presents a service that takes as
input many individual, noisy heuristics over an unlabeled data set, and
produces combined labels for the data as well as an assessment of which
heuristics were most accurate for which classes. What really intrigued
me is that the system works even for _partial_ heuristics—i.e.,
heuristics that abstain from labeling some of the data points—which has
been a critical problem in our previous explorations (see discussions
above).

One potential sticking point is that it may not be obvious how to apply
Snorkel to continuous label spaces, which cred must surely be. But,
imho, this is interesting enough that we should at least take a good
look at it. @000drax also speaks highly of the people behind the
project, having worked with them on an application and paper:
<https://arxiv.org/abs/1812.00417>

**Comment by ajratner** [comment]
@wchargin thanks for checking out the project!  And glad you note the utility of the abstentions- turns out, though its a simple addition to the semantics, it is often quite useful :) . And glad to have the excuse to check out SourceCred- looks really cool!  I especially like the core design around leaving flexibility in the design of the heuristics, edge types / weights, etc, then using a standard and thus transparent approach to compute the scores ultimately.

We have had some ideas of extending Snorkel to continuous spaces.  To start, are your continuous spaces normalized?  E.g. can I think of them all in [0,1]?

**Comment by Beanow** [comment]
With the `WeightedGraph` type landing (#1590). Is this still an open discussion?

**Comment by decentralion** [comment]
This isn't resolved by #1590, but it is quite old and our thinking on weights (and the constraints around weights compositions) has shifted a lot since it was written. Therefore, I'll close it.
